keys.go

```diff
@@ -0,0 +1,109 @@
+package main
+
+import (
+	"bytes"
+	"fmt"
+	"os"
+	"os/exec"
+	"strings"
+
+	gopkg "github.com/mikkeloscar/gopkgbuild"
+)
+
+// pgpKeySet maps a PGP key with a list of PKGBUILDs that require it.
+// This is similar to stringSet, used throughout the code.
+type pgpKeySet map[string][]string
+
+func (set pgpKeySet) toSlice() []string {
+	slice := make([]string, 0, len(set))
+	for v := range set {
+		slice = append(slice, v)
+	}
+	return slice
+}
+
+func (set pgpKeySet) set(key, pkgbasename string) {
+	// Using ToUpper to make sure keys with a different case will be
+	// considered the same.
+	upperKey := strings.ToUpper(key)
+	if _, exists := set[upperKey]; !exists {
+		set[upperKey] = []string{}
+	}
+	set[key] = append(set[key], pkgbasename)
+}
+
+// checkPgpKeys iterates through the keys listed in the PKGBUILDs and if needed,
+// asks the user whether yay should try to import them. gpgExtraArgs are extra
+// parameters to pass to gpg, in order to facilitate testing, such as using a
+// different keyring. It can be nil.
+func checkPgpKeys(gpgExtraArgs []string, srcinfos map[string]*gopkg.PKGBUILD) error {
+	// keys maps a given PGP with boolean to indicate whether we would need to
+	// import this particular key.
+	keys := make(pgpKeySet)
+
+	// Mapping all the keys.
+	for _, pkgbuild := range srcinfos {
+		for _, key := range pkgbuild.Validpgpkeys {
+			keys.set(key, pkgbuild.Pkgbase)
+		}
+	}
+
+	// Let's check the keys individually, and then we can offer to import
+	// the problematic ones.
+	problematic := make(pgpKeySet)
+	args := append(gpgExtraArgs, "--list-keys")
+	for key := range keys {
+		cmd := exec.Command(config.GpgBin, append(args, key)...)
+		err := cmd.Run()
+		if err != nil {
+			problematic[key] = keys[key]
```

Comment:
Just use one pgpKeySet and `delete(keys, key)` if it's not problematic.

Reply:
Or actually just do the opposite and only add problematic keys in the first place.